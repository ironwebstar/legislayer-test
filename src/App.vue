<template>
  <div id="app">
    <router-view />
  </div>
</template>

<style lang="scss">
@import "./assets/theme/variables";

#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: $color-black;
  font-size: $font-normal;
}
</style>
