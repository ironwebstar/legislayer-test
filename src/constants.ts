export const SELECTED_USER = "SELECTED_USER";

export const PROFILES = "PROFILES";

export const LOADING_ITEMS = 25;
