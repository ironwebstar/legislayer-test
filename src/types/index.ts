export interface Profile {
  id: {
    name: string;
    value: string | null;
  };
  gender: string;
  name: {
    title: string;
    first: string;
    last: string;
  };
  location: {
    street: {
      name: string;
      number: number;
    };
    city: string;
    country: string;
    state: string;
    postcode: number | string;
    coordinates: {
      latitude: string;
      longitude: string;
    };
    timezone: {
      description: string;
      offset: string;
    };
  };
  phone: string;
  cell: string;
  picture: {
    large: string;
    medium: string;
    thumbnail: string;
  };
}
