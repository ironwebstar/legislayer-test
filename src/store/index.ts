import Vue from "vue";
import Vuex from "vuex";
import { MutationTree, GetterTree, ActionTree, ActionContext } from "vuex";

import { getProfiles } from "../util/rest";
// import { mergeProfiles } from "../util/transforms";
import { Profile } from "../types";
import { LOADING_ITEMS, PROFILES, SELECTED_USER } from "../constants";

Vue.use(Vuex);

export interface RootState {
  profiles: Profile[];
  curProfile: Profile | null;
}
interface RootStateFactory {
  (): RootState;
}

export const rootState: RootStateFactory = () => ({
  profiles: [],
  curProfile: null,
});

export const actions: ActionTree<RootState, any> = {
  async LOAD_PROFILES(
    ctx: ActionContext<RootState, any>,
    payload: {
      results?: number;
      gender?: boolean;
    }
  ) {
    const { results, gender } = payload;
    const loadingResults = results || LOADING_ITEMS;

    try {
      const profiles = await getProfiles({ results: loadingResults, gender });
      // const curProfiles = mergeProfiles(profiles.results, ctx.state.profiles);

      ctx.commit("SET_PROFILES", [...ctx.state.profiles, ...profiles.results]);
    } catch (error) {
      if (process.env.NODE_ENV !== "production") {
        console.error("ERROR: NO PROFILE DATA");
      }
      return { error };
    }
  },
};

export const mutations: MutationTree<RootState> = {
  SET_PROFILES(state: RootState, profiles: Profile[]) {
    state.profiles = [...profiles];
    console.log(profiles);
    localStorage.setItem(PROFILES, JSON.stringify(state.profiles));
  },

  SET_CUR_PROFILE(state: RootState, curProfile: Profile) {
    state.curProfile = curProfile;
    localStorage.setItem(SELECTED_USER, JSON.stringify(curProfile));
  },
};

export const getters: GetterTree<RootState, any> = {
  profiles: (state: RootState) => state.profiles,
  curProfile: (state: RootState) => state.curProfile,
};

export default new Vuex.Store({
  state: rootState,
  mutations,
  getters,
  actions,
});
