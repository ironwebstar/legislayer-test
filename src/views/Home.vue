<template>
  <div class="home">
    <div class="search">
      <div class="search-input">
        <input class="" v-model="search" @input="searchHandle" />
      </div>

      <div class="search-list">
        <profile-item
          v-for="(profile, index) in filteredProfiles"
          :key="index"
          :profile="profile"
          @click.native="selectProfile(profile)"
        />
        <infinite-loading
          :identifier="infiniteId"
          @infinite="infiniteHandler"
        ></infinite-loading>
      </div>
    </div>

    <div class="profile">
      <div v-if="curProfile">
        <div class="current-profile">
          <div class="current-profile-top">
            <avatar
              :image="curProfile.picture.thumbnail"
              :width="100"
              :height="100"
            />
            <div class="current-profile-top-content">
              <div class="current-profile-name">
                {{
                  `${curProfile.name.title}. ${curProfile.name.first} ${curProfile.name.last}`
                }}
              </div>
              <p class="current-profile-username">
                {{ `@${curProfile.login.username}` }}
              </p>
              <a
                class="current-profile-email"
                :href="`mailto:${curProfile.email}`"
              >
                Email: {{ curProfile.email }}
              </a>
              <a class="current-profile-cell" :href="`tel:${curProfile.cell}`">
                Phone: {{ curProfile.cell }}
              </a>
            </div>
          </div>

          <div class="current-profile-bottom">
            <p><b>Location:</b> {{ profileLocation }}</p>
            <p><b>Timezone:</b> {{ profileTimezone }}</p>
            <p><b>Birthday:</b> {{ curProfile.dob.age }}</p>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import debounce from "debounce";
import { Getter, Mutation, Action } from "vuex-class";
import Component from "vue-class-component";
import { Watch } from "vue-property-decorator";
import InfiniteLoading from "vue-infinite-loading";

import ProfileItem from "../components/ProfileItem.vue";
import Avatar from "../components/Avatar.vue";

import { Profile } from "../types";
import { SELECTED_USER, LOADING_ITEMS } from "../constants";

@Component({
  components: { ProfileItem, Avatar, InfiniteLoading },
})
export default class Home extends Vue {
  @Getter("profiles") profiles!: Profile[];
  @Getter("curProfile") curProfile!: Profile;

  @Mutation("SET_PROFILES") setProfiles: any;
  @Mutation("SET_CUR_PROFILE") setCurProfile: any;
  @Action("LOAD_PROFILES") loadProfiles: any;

  search = "";
  filteredProfiles: Profile[] = [];
  infiniteId = +new Date();

  @Watch("profiles")
  onProfilesChange(): void {
    this.filteredProfiles = this.profiles;
  }

  created(): void {
    this.filteredProfiles = this.profiles;
    this.setCurProfile(
      localStorage.getItem(SELECTED_USER)
        ? JSON.parse(localStorage.getItem(SELECTED_USER) || "")
        : null
    );
  }

  infiniteHandler(): void {
    this.loadProfiles({ results: LOADING_ITEMS });
  }

  async searchHandle(): Promise<void> {
    this.setProfiles([]);
    debounce(async () => {
      if (this.search) await this.searchProfile();
    }, 500);
  }

  async searchProfile(): Promise<void> {
    const search = this.search.toLowerCase();

    if (this.profiles.length < LOADING_ITEMS) {
      await this.loadProfiles({ results: LOADING_ITEMS });
    }

    if (search) {
      const filtered = await this.profiles.filter(
        (profile: Profile) =>
          profile.name.first.toLowerCase().includes(search) ||
          profile.name.last.toLowerCase().includes(search)
      );
      if (filtered.length > LOADING_ITEMS) filtered.length = LOADING_ITEMS;

      this.filteredProfiles = filtered;
    }
  }

  selectProfile(profile: Profile): void {
    this.setCurProfile(profile);
  }

  get profileLocation(): string {
    if (this.curProfile) {
      const { location } = this.curProfile;
      return `${location.street.name} ${location.street.number}, ${location.city}, ${location.state}, ${location.country}`;
    }
    return "";
  }

  get profileTimezone(): string {
    if (this.curProfile) {
      const { location } = this.curProfile;
      return `${location.timezone.description} ${location.timezone.offset}`;
    }
    return "";
  }
}
</script>

<style lang="scss" scoped>
@import "../assets/theme/variables";

.home {
  margin: $spacing * 2 auto;
  display: flex;
  width: 90%;
}

.search {
  width: 25%;
  display: flex;
  margin: 0 auto;
  flex-direction: column;

  &-input {
    width: 100%;

    input {
      width: 90%;
      padding: $spacing * 0.5;
      margin: 0 auto;
    }
  }

  &-list {
    display: flex;
    flex-direction: column;
    padding: $spacing 0;
  }
}

.profile {
  width: 75%;
  padding: $spacing $spacing * 2;
  text-align: left;
}

.current-profile {
  &-top {
    display: flex;
    width: 100%;

    &-content {
      padding-left: $spacing * 3;
      display: flex;
      flex-direction: column;
    }
  }

  &-name {
    font-weight: bold;
    font-size: $font-big;
    padding-bottom: $spacing * 0.5;
  }

  &-bottom {
    padding: $spacing * 2 0;
    p {
      display: flex;
      flex-direction: column;
      margin-bottom: $spacing;
    }
  }
}
</style>
