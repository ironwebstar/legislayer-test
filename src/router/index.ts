import Vue from "vue";
import VueRouter, { RouteConfig } from "vue-router";

import Home from "../views/Home.vue";
import Store from "../store";
import { PROFILES, LOADING_ITEMS } from "../constants";

Vue.use(VueRouter);

const routes: Array<RouteConfig> = [
  {
    path: "/",
    name: "Home",
    component: Home,
    async beforeEnter(to, from, next) {
      try {
        if (!localStorage.getItem(PROFILES))
          await Store.dispatch("LOAD_PROFILES", { results: LOADING_ITEMS });
        else {
          const localProfiles = localStorage.getItem(PROFILES)
            ? JSON.parse(localStorage.getItem(PROFILES) || "")
            : [];
          if (localProfiles.length > LOADING_ITEMS)
            localProfiles.length = LOADING_ITEMS;
          Store.commit("SET_PROFILES", localProfiles);
        }
        next();
      } catch (error) {
        // TODO: Error Handling
      }
    },
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

export default router;
