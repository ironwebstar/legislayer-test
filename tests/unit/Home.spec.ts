import { createLocalVue, shallowMount, Wrapper } from "@vue/test-utils";
import Vuex, { GetterTree, MutationTree, ActionTree } from "vuex";

import Home from "../../src/views/Home.vue";
import { profiles } from "../mock-data/profiles";
import { RootState } from "../../src/store";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("Home", () => {
  let actions: ActionTree<RootState, any>;
  let state: RootState;
  let store;
  let getters: GetterTree<RootState, any>;
  let mutations: MutationTree<RootState>;

  let wrapper: Wrapper<any>;
  beforeEach(() => {
    state = {
      profiles,
      curProfile: null,
    };
    actions = {
      LOAD_PROFILES: jest.fn(),
    };

    mutations = {
      SET_CUR_PROFILE: () => jest.fn(),
    };

    getters = {
      profiles: () => profiles,
      curProfile: () => profiles[0],
    };

    store = new Vuex.Store({
      state,
      actions,
      getters,
      mutations,
    });

    wrapper = shallowMount(Home, { store, localVue });
  });

  it("snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("get profile location if current profile is existed", () => {
    expect(wrapper.vm.profileLocation).toBe(
      "Mill Lane 2958, Swords, Galway, Ireland"
    );
    expect(wrapper.vm.profileTimezone).toBe("Almaty, Dhaka, Colombo +6:00");
  });

  it("get profile location if current profile is not existed", async () => {
    store = new Vuex.Store({
      state,
      actions,
      getters: {
        ...getters,
        curProfile: () => null,
      },
      mutations,
    });
    wrapper = shallowMount(Home, { store, localVue });
    expect(wrapper.vm.profileLocation).toBe("");
    expect(wrapper.vm.profileTimezone).toBe("");
  });

  it("select profile", async () => {
    await wrapper.vm.selectProfile(profiles[0]);
    expect(wrapper.vm.curProfile).toBe(profiles[0]);
  });
});
