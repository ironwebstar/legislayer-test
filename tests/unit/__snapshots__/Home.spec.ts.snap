// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot 1`] = `
"<div class=\\"home\\">
  <div class=\\"search\\">
    <div class=\\"search-input\\"><input></div>
    <div class=\\"search-list\\">
      <profile-item-stub profile=\\"[object Object]\\"></profile-item-stub>
      <profile-item-stub profile=\\"[object Object]\\"></profile-item-stub>
      <profile-item-stub profile=\\"[object Object]\\"></profile-item-stub>
      <profile-item-stub profile=\\"[object Object]\\"></profile-item-stub>
      <profile-item-stub profile=\\"[object Object]\\"></profile-item-stub>
      <infinite-loading-stub distance=\\"100\\" direction=\\"bottom\\" identifier=\\"1626681263264\\"></infinite-loading-stub>
    </div>
  </div>
  <div class=\\"profile\\">
    <div>
      <div class=\\"current-profile\\">
        <div class=\\"current-profile-top\\">
          <avatar-stub image=\\"https://randomuser.me/api/portraits/thumb/men/2.jpg\\" width=\\"100\\" height=\\"100\\"></avatar-stub>
          <div class=\\"current-profile-top-content\\">
            <div class=\\"current-profile-name\\">
              Mr. Herman Warren
            </div>
            <p class=\\"current-profile-username\\">
              @heavyelephant650
            </p> <a href=\\"mailto:herman.warren@example.com\\" class=\\"current-profile-email\\">
              Email: herman.warren@example.com
            </a> <a href=\\"tel:081-891-1455\\" class=\\"current-profile-cell\\">
              Phone: 081-891-1455
            </a>
          </div>
        </div>
        <div class=\\"current-profile-bottom\\">
          <p><b>Location:</b> Mill Lane 2958, Swords, Galway, Ireland</p>
          <p><b>Timezone:</b> Almaty, Dhaka, Colombo +6:00</p>
          <p><b>Birthday:</b> 24</p>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
